src/aspire/source/__init__.py

```diff
@@ -370,6 +372,108 @@ def whiten(self, noise_filter):
         # Invalidate images
         self._im = None
 
+    def phase_flip(self):
+        """
+        Perform phase flip to images in the source object using CTF information
+        """
+        logger.info('Perform phase flip on source object')
+        logger.info('Adding Phase Flip Xform to end of generation pipeline')
+        self.generation_pipeline.add_xform(FlipXform(self.filters))
+        # Invalidate images
+        self._im = None
+
+    def reverse_contrast(self, batch_size=512):
+        """
+        Reverse the global contrast of images
+
+        Check if all images in a stack should be globally phase flipped so that
+        the molecule corresponds to brighter pixels and the background corresponds
+        to darker pixels. This is done by comparing the mean in a small circle
+        around the origin (supposed to correspond to the molecule) with the mean
+        of the noise, and making sure that the mean of the molecule is larger.
+        From the implementation level, we modify the `ImageSource` in-place by
+        appending a `Multiple` filter to the generation pipeline.
+
+        :param batch_size: Batch size of images to query.
+        :return: On return, the `ImageSource` object has been modified in place.
+        """
+
+        logger.info('Apply contrast reverse on source object')
+        L = self.L
+        image_center = (L + 1) / 2
+        coor_mat_m, coor_mat_n = np.meshgrid(np.arange(1, L + 1), np.arange(1, L + 1))
+
+        distance_from_center = np.sqrt((coor_mat_m - image_center) ** 2
```

Comment:
Why not use `grid_2d` as in `normalize_background` below?

Reply:
Done in commit 9e0ba0e.